File: sdks/typescript/test/conformance.test.js
import test from "node:test";
import assert from "node:assert/strict";
import { readFile } from "node:fs/promises";
import {
  NexaV6Client,
  canonicalJson,
  computeFeedHash,
  requestPermitMessage,
  SDK_HEADER_VALUE,
  verifyFeed,
} from "../src/index.js";

const vectors = JSON.parse(await readFile(
  new URL("../../../sdk-spec/test-vectors.json", import.meta.url),
  "utf8",
));

test("canonical JSON and Feed verification match frozen vectors", () => {
  assert.equal(
    canonicalJson(vectors.canonicalJson.input),
    vectors.canonicalJson.expected,
  );
  assert.equal(
    computeFeedHash(vectors.feed.signedPayload),
    vectors.feed.feedHash,
  );
  const result = verifyFeed({
    signedPayload: vectors.feed.signedPayload,
    feedHash: vectors.feed.feedHash,
    feedSigner: vectors.feed.feedSigner,
    feedSignature: vectors.feed.feedSignature,
  }, {
    expectedSigner: vectors.feed.feedSigner,
    nowSeconds: vectors.feed.nowSeconds,
    required: true,
  });
  assert.equal(result.valid, true);
  assert.equal(result.recoveredSigner, vectors.feed.feedSigner);
});

test("Permit message and execution calldata match frozen vectors", () => {
  assert.equal(
    requestPermitMessage(vectors.permitRequest.request),
    vectors.permitRequest.expectedMessage,
  );
  const client = new NexaV6Client({ fetch: async () => new Response("{}") });
  const transaction = client.buildExecutionTx({
    permit: vectors.abi.permit,
    permitSignature: vectors.abi.permitSignature,
    execution: { target: vectors.abi.executionTarget },
  });
  assert.equal(transaction.data, vectors.abi.fillDirectCallData);
  assert.equal(transaction.value, vectors.abi.expectedTransactionValue);
});

test("all frozen behavioral operations are present", () => {
  const client = new NexaV6Client({ fetch: async () => new Response("{}") });
  for (const name of [
    "discover", "getRoutes", "getRoute", "verifyFeed",
    "requestPermitMessage", "requestPermit", "resolveExecution",
    "previewExecution", "buildExecutionTx", "getFillStatus",
  ]) {
    assert.equal(typeof client[name], "function", name);
  }
});

test("public Nexa requests declare the SDK but RPC requests do not", async () => {
  const requests = [];
  const client = new NexaV6Client({
    fetch: async (_url, init) => {
      requests.push(init);
      return new Response(JSON.stringify({
        schema: "NEXA_MAINNET_V6_SOLVER_DISCOVERY_V2",
        deploymentVersion: 6,
        deploymentStatus: "ACTIVE",
        releaseId: `0x${"11".repeat(32)}`,
        feedSigner: "0x0000000000000000000000000000000000000001",
        endpoints: { solverFeed: "https://solver.vsnexa.com/api/v6/solver-feed" },
      }), { headers: { "content-type": "application/json" } });
    },
  });
  await client.discover();
  await assert.rejects(
    () => client.rpc("https://rpc.example", "eth_call", []),
    /NEXA_SDK_RPC_ERROR/,
  );
  assert.equal(new Headers(requests[0].headers).get("x-nexa-sdk"), SDK_HEADER_VALUE);
  assert.equal(new Headers(requests[1].headers).get("x-nexa-sdk"), null);
});

File: sdks/jvm/src/test/java/com/vsnexa/v6/ConformanceTest.java
package com.vsnexa.v6;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.fasterxml.jackson.databind.JsonNode;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Set;
import java.util.stream.Collectors;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

final class ConformanceTest {
  private static JsonNode vectors;

  @BeforeAll
  static void loadVectors() throws Exception {
    vectors = NexaV6Core.JSON.readTree(Files.readString(
        Path.of("..", "..", "sdk-spec", "test-vectors.json")));
  }

  @Test
  void canonicalFeedAndSignatureMatchFrozenVectors() {
    assertEquals(vectors.path("canonicalJson").path("expected").asText(),
        NexaV6Core.canonicalJson(vectors.path("canonicalJson").path("input")));
    JsonNode feedVector = vectors.path("feed");
    assertEquals(feedVector.path("feedHash").asText(),
        NexaV6Core.computeFeedHash(feedVector.path("signedPayload")));
    var feed = NexaV6Core.JSON.createObjectNode();
    feed.set("signedPayload", feedVector.path("signedPayload"));
    feed.set("feedHash", feedVector.path("feedHash"));
    feed.set("feedSigner", feedVector.path("feedSigner"));
    feed.set("feedSignature", feedVector.path("feedSignature"));
    var result = NexaV6Core.verifyFeed(feed, feedVector.path("feedSigner").asText(),
        feedVector.path("nowSeconds").asLong(), true);
    assertTrue(result.valid());
    assertEquals(feedVector.path("feedSigner").asText(), result.recoveredSigner());
  }

  @Test
  void permitMessageAndCalldataMatchFrozenVectors() {
    JsonNode permitVector = vectors.path("permitRequest");
    assertEquals(permitVector.path("expectedMessage").asText(),
        NexaV6Core.requestPermitMessage(permitVector.path("request")));
    JsonNode abi = vectors.path("abi");
    var envelope = NexaV6Core.JSON.createObjectNode();
    envelope.set("permit", abi.path("permit"));
    envelope.set("permitSignature", abi.path("permitSignature"));
    var execution = envelope.putObject("execution");
    execution.set("target", abi.path("executionTarget"));
    var transaction = NexaV6Abi.buildExecutionTx(envelope);
    assertEquals(abi.path("fillDirectCallData").asText(), transaction.data());
    assertEquals(abi.path("expectedTransactionValue").asText(), transaction.value());
  }

  @Test
  void allFrozenOperationsArePresent() {
    Set<String> methods = java.util.Arrays.stream(NexaV6Client.class.getDeclaredMethods())
        .map(java.lang.reflect.Method::getName).collect(Collectors.toSet());
    for (String name : Set.of("discover", "getRoutes", "getRoute", "verifyFeed",
        "requestPermitMessage", "requestPermit", "resolveExecution", "previewExecution",
        "buildExecutionTx", "getFillStatus")) {
      assertTrue(methods.contains(name), name);
    }
  }
}
